Refactoring and cleanup of `species.param`

steindev opened the pull request:
Actions performed:
* Rename particle shapes to be consistent in naming -> requires renaming of the respective `*.hpp` files and classes therein
* Apply these changes to examples and test
* Add `getStringProperty()` to `FieldToParticleInterpolationNative` class to make it compilable, i.e. usable
* Clean examples from `species.param` files that were just old standards and never updated -> this actually means that most examples use `particles::shapes::TSC` instead of `particles::shapes::CIC` now

sbastrakov commented on include/picongpu/param/species.param:
```diff
@@ -38,12 +38,23 @@ namespace picongpu
+ *      -> PCS is the name of the 4th order shape (instead of P4S)
+ *      -> P4S does not exist anymore
+ *  **
+ *  - particles::shapes::CIC : 1st order, assignment function is a piecewise linear spline
+ *  - particles::shapes::TSC : 2nd order, assignment function is a piecewise quadratic spline
+ *  - particles::shapes::PQS : 3rd order, assignment function is a piecewise cubic spline
+ *  - particles::shapes::PCS : 4th order, assignment function is a piecewise quartic spline
```
What does the `order` part of the comment add? The information on the assignment function is given after the comma. While out of context this `order` may be misunderstood as the cloud density order.

steindev replied:
Well, nothing anymore. Just didn't remove it. Will do now.

sbastrakov commented on share/picongpu/tests/compileParticlePusher/include/picongpu/param/species.param:
```diff
@@ -17,54 +17,74 @@
+ *  Particle shape naming is according to
+ *    Hockney R, Eastwood JW. Computer simulation using particles. IOP Pubslishing Ltd (1988). p. 144.
+ *  That is, the name of a particle shape is defined by the visual form of its cloud shape S,
+ *  which describes a particle's charge density distribution.
```
Macroparticle's ?

sbastrakov commented on share/picongpu/tests/compileParticlePusher/include/picongpu/param/species.param:
```diff
@@ -17,54 +17,74 @@
+ *  However, the charge assigned to a mesh point due to a distribution of overlapping finite-sized
+ *  particles is obtained by evaluating and summing the assignment function W for all particles.
+ *  The assignment function is obtained by convolving the cloud shape with the top-hat function.
+ *  PIConGPU only implements the assignment functions.
```
I am not sure this information is relevant at this point. I would imagine a user cares about how to set the desired properties, not how it's implemented internally

psychocoderHPC replied:
I would also move such information to the documentation instead of having it here.
Should we also shorten this section and move it to the documentation and link the documentation here:

```
...

**WARNING** the shape naming has changed and is not consistent too PIConGPU versions before 0.6.0. see readthedocslink

  - particles::shapes::CIC : 1st order, assignment function is a piecewise linear spline
  - particles::shapes::TSC : 2nd order, assignment function is a piecewise quadratic spline
  - particles::shapes::PQS : 3rd order, assignment function is a piecewise cubic spline
  - particles::shapes::PCS : 4th order, assignment function is a piecewise quartic spline
```

steindev replied:
I moved most of the text to the general file documentation, where I think it belongs and is visible at readthedocs. I also think we can leave the explaining sentence on PIConGPU imlementing only assignment functions at this general file documentation level, as it will help users/developers understanding what PIConGPU does. I agree, it is superfluous as a comment to `using UsedParticleShape`.

sbastrakov commented on share/picongpu/tests/compileParticlePusher/include/picongpu/param/species.param:
```diff
@@ -17,54 +17,74 @@
+/** define which interpolation method is used to interpolate fields to particles
+ */
+using UsedField2Particle = FieldToParticleInterpolation<
+    UsedParticleShape,
+    AssignedTrilinearInterpolation
+>;
 
 /*! select current solver method
- * - currentSolver::Esirkepov< SHAPE, STRATEGY > : particle shapes - CIC, TSC, PCS, P4S (1st to 4th order)
+ * - currentSolver::Esirkepov< SHAPE, STRATEGY > : particle shapes - CIC, TSC, PQS, PCS (1st to 4th order)
  * - currentSolver::VillaBune< SHAPE, STRATEGY > : particle shapes - CIC (1st order) only
- * - currentSolver::EmZ< SHAPE, STRATEGY >       : particle shapes - CIC, TSC, PCS, P4S (1st to 4th order)
+ * - currentSolver::EmZ< SHAPE, STRATEGY >       : particle shapes - CIC, TSC, PQS, PCS (1st to 4th order)
  *
  * For development purposes:
- * - currentSolver::currentSolver::EsirkepovNative< SHAPE, STRATEGY > : generic version of currentSolverEsirkepov
+ * - currentSolver::EsirkepovNative< SHAPE, STRATEGY > : generic version of currentSolverEsirkepov
  *   without optimization (~4x slower and needs more shared memory)
  *
  * STRATEGY (optional):
- * - currentSolver::strategy::StridedCachedSupercells
- * - currentSolver::strategy::CachedSupercells
- * - currentSolver::strategy::NonCachedSupercells
+ * - currentSolver::StridedCachedSupercells
+ * - currentSolver::CachedSupercells
+ * - currentSolver::NonCachedSupercells
```
Not sure if this is correct. That `strategy` namespace seems still in place

steindev replied:
Hmm that happened accidentally when replacing the old file in the test by the new one. As this usage is explained in the standard `species.param`, I assume it is the recommended way.
And that the documentation in this test's `species.param` has never been updated but reflects an old standard.

psychocoderHPC replied:
`currentSolver::strategy::` must be included!

steindev replied:
:ok:

sbastrakov commented:
@steindev thanks, I like these changes in principle, some small notes above. However, why do they all have to be in the same PR? To me it seems like the changes are of 3 kinds:

- Renaming of shapes and updating comments on that
- Other cleaup in species.param
- Changes to `FieldToParticleInterpolationNative `

Couldn't they all be separate PRs ?

steindev commented:
@sbastrakov and @psychocoderHPC: I updated the documentation in `species.param` according to your comments. Please have a look. If you think it is ok now, I will also update all the other `species.param` files in the examples and push again.

steindev commented:
> However, why do they all have to be in the same PR? [...]
> Couldn't they all be separate PRs ?

Well I grouped in 4 different commits, so they can end up in PIConGPU as 4 different commits, which will be the same result as using individual PRs. In the end, it will therefore make no difference. But as all these logically belong together and are in itself just small changes, I thought it would be sufficient to have only on PR for these and thereby reduce the additional effort.

PrometheusPi commented on share/picongpu/tests/XrayScattering/include/picongpu/param/species.param:
```diff
@@ -1,4 +1,4 @@
-/* Copyright 2014-2020 Rene Widera, Richard Pausch, Pawel Ordyna
```
why is Annegret listed here as co-author 
(applies to various other files as well)

steindev replied:
She implemented the Higuera-Cary particle pusher which can be selected in the `species.param`. Since she edited this file to make it available she is an author. And since I copied the new standard file to all examples, she is now an author in several files were she hasn't been an author before.

PrometheusPi reviewed (approved):
Looks good to me, just a question regarding multiple new authors in various files.

sbastrakov commented on include/picongpu/param/species.param:
```diff
@@ -19,9 +19,25 @@
+ * Macroparticle shape naming is according to
+ *    Hockney R, Eastwood JW. Computer simulation using particles. IOP Pubslishing Ltd (1988). p. 144.
+ * 
+ * The name of macroparticle shape is defined by the visual form of its cloud shape S,
+ * which describes a macroparticle's charge density distribution.
+ * However, the charge assigned to a mesh point due to a distribution of overlapping finite-sized
+ * macroparticles is obtained by evaluating and summing the assignment function W for all particles.
+ * The assignment function is obtained by convolving the cloud shape with the top-hat function.
+ * PIConGPU only implements the assignment functions.
+ *
+ * @warning
+ * The higher order shape names are redefined with release 0.6.0 in order to provide a consistent naming:
+ *     -> PQS is the name of the 3rd order shape (instead of PCS)
+ *     -> PCS is the name of the 4th order shape (instead of P4S)
+ *     -> P4S does not exist anymore
```
I think this piece is not self-consistent. From the first sentence a reader may assume everything is about cloud shapes. But these `3rd` and `4th order shape` are **not** shape orders. Rather, they are assignment function orders

sbastrakov replied:
I think the text should be corrected to have `3rd order assignment function` instead of `3rd order shape`

steindev commented:
@sbastrakov, @psychocoderHPC: I expanded the documentation and again updated `species.param` accordingly. As before, I still need to update all other `species.param`s which I will do after your review.

psychocoderHPC commented on share/picongpu/examples/SingleParticleTest/include/picongpu/param/species.param:
```diff
@@ -17,51 +17,71 @@
- * - currentSolver::strategy::StridedCachedSupercells
- * - currentSolver::strategy::CachedSupercells
- * - currentSolver::strategy::NonCachedSupercells
+ * - currentSolver::StridedCachedSupercells
```
Please revert this change it must be `currentSolver::strategy::StridedCachedSupercells`

sbastrakov replied:
Wait, I was already asking about it for the very first version of this PR.

steindev commented:
@psychocoderHPC I addressed your comments and included the refactored `species.param` into all examples and tests. CIs are happy, too. Ready to merge?!

PrometheusPi commented on docs/source/models/pic.rst:
```diff
@@ -64,20 +64,12 @@ Electro-Magnetic PIC Method
-The three dimensional marker shape is a multiplicative union of B-splines :math:`S^n(x,y,z) = S^n(x) S^n(y) S^n(z)`.
+A marker has a finite-size and a velocity, such that it can be regarded as a cloud of particles, whose center of mass is the marker's position and whose mean velocity is the marker's velocity.
+The cloud shape :math:`S^n(x)` of order :math:`n` of a marker describes its charge density distribution.
+See :ref:`section Hierarchy of charge assignment schemes <model-shapes>` for a list of available marker shapes in PIConGPU.
```
typo: `Hierarchy` -> `hierarchy`

steindev replied:
I will correct to the correct title with capital nouns.

PrometheusPi commented on docs/source/models/shapes.rst:
```diff
@@ -0,0 +1,64 @@
+            0 & |x| \gt \frac{1}{2} \\
+            \frac{1}{2} & |x| = \frac{1}{2} \\
+            1 & |x| \lt \frac{1}{2}
+      \end{array}\right.\,.
```
Why do you end the equation with a dot? 
It is not the end of a sentence.

PrometheusPi commented on docs/source/models/shapes.rst:
```diff
@@ -0,0 +1,64 @@
+The assignment function of order :math:`n` is generated by convolution of the assignment function of order :math:`n-1` with the top-hat function
+
+.. math::
+
```
does this space breaks the math mode?

steindev replied:
No.

PrometheusPi commented on share/picongpu/tests/XrayScattering/include/picongpu/param/species.param:
```diff
@@ -1,4 +1,4 @@
-/* Copyright 2014-2020 Rene Widera, Richard Pausch, Pawel Ordyna
+/* Copyright 2014-2020 Rene Widera, Richard Pausch, Annegret Roeszler, Klaus Steiniger
```
You accidentally removed @pordyna

PrometheusPi commented on share/picongpu/tests/compileParticlePusher/include/picongpu/param/species.param:
```diff
@@ -17,54 +17,74 @@
  * - currentSolver::strategy::NonCachedSupercells
  */
-using UsedParticleCurrentSolver = currentSolver::EmZ<UsedParticleShape>;
+using UsedParticleCurrentSolver = currentSolver::Esirkepov< UsedParticleShape >;
```
Why not EmZ anymore?

steindev replied:
My fault...

steindev commented on share/picongpu/tests/compileParticlePusher/include/picongpu/param/species.param:
```diff
@@ -1,4 +1,4 @@
-/* Copyright 2014-2020 Rene Widera, Richard Pausch
+/* Copyright 2014-2020 Rene Widera, Richard Pausch, Annegret Roeszler, Klaus Steiniger
```
Add sbastrakov

psychocoderHPC commented:
grrrr :angry: Searched over 2h why my updated HIP branch is now slower than the old code base. After a diff I realiced tha twe change the shape names and I simulated with the shape `PCS` which has now a new order.
@jkelling with the next SPEC benchmark setup update we need to change https://github.com/psychocoderHPC/picongpu/blob/f22742b9fbfad7a5d725b41727991f1f63c01b00/share/picongpu/examples/SPECBenchmark/include/picongpu/param/species.param#L48 to `PQS`

steindev commented:
:rofl: Did you use the TWEAC setup for testing, not any standard example? Wouldn't have expected that any test setup uses the higher order shapes...

But 2h is a significant loss these days :grimacing:

psychocoderHPC commented:
> 🤣 Did you use the TWEAC setup for testing, not any standard example? Wouldn't have expected that any test setup uses the higher order shapes...
> 
> But 2h is a significant loss these days 😬

Yes the SPEC benchmark is currently not part of the examples. The advantage is that it requires only single precision and the amount of memory is scaling linear with the volume per GPU. There is one update required before I can add this example to our codebase.
